File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/atl/PopupContent.java
package com.adadapted.android.sdk.core.atl;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class PopupContent implements AddToListContent, Parcelable {
    private final String payloadId;
    private final List<AddToListItem> items;

    private boolean handled;
    private final Lock lock = new ReentrantLock();

    public static PopupContent createPopupContent(final String payloadId,
                                                  final List<AddToListItem> items) {
        return new PopupContent(payloadId, items);
    }

    public PopupContent(final String payloadId, final List<AddToListItem> items) {
        this.payloadId = payloadId;
        this.items = items;
    }

    protected PopupContent(Parcel in) {
        payloadId = in.readString();
        items = in.createTypedArrayList(AddToListItem.CREATOR);
        handled = in.readByte() != 0;
    }

    public static final Creator<PopupContent> CREATOR = new Creator<PopupContent>() {
        @Override
        public PopupContent createFromParcel(Parcel in) {
            return new PopupContent(in);
        }

        @Override
        public PopupContent[] newArray(int size) {
            return new PopupContent[size];
        }
    };

    @Override
    public void acknowledge() {
        lock.lock();
        try {
            if (!handled) {
                handled = true;
                PopupClient.markPopupContentAcknowledged(this);
            }
        }
        finally {
            lock.unlock();
        }
    }

    @Override
    public void itemAcknowledge(final AddToListItem item) {
        lock.lock();
        try {
            if (!handled) {
                handled = true;
                PopupClient.markPopupContentAcknowledged(this);
            }

            PopupClient.markPopupContentItemAcknowledged(this, item);
        }
        finally {
            lock.unlock();
        }
    }

    @Override
    public void failed(final String message) {
        lock.lock();
        try {
            if (!handled) {
                handled = true;
                PopupClient.markPopupContentFailed(this, message);
            }
        }
        finally {
            lock.unlock();
        }
    }

    @Override
    public void itemFailed(final AddToListItem item, final String message) {
        lock.lock();
        try {
            PopupClient.markPopupContentItemFailed(this, item, message);
        }
        finally {
            lock.unlock();
        }
    }

    @Override
    public String getSource() {
        return Sources.IN_APP;
    }

    public String getPayloadId() {
        return payloadId;
    }

    @Override
    public List<AddToListItem> getItems() {
        return items;
    }

    @Override
    public boolean hasNoItems() {
        return items.size() == 0;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(payloadId);
        parcel.writeTypedList(items);
        parcel.writeByte((byte) (handled ? 1 : 0));
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/atl/PopupContent.kt
package com.adadapted.android.sdk.core.atl

import com.adadapted.android.sdk.constants.EventStrings
import com.adadapted.android.sdk.core.event.EventClient

class PopupContent(val payloadId: String, private val items: List<AddToListItem>) :
    AddToListContent {
    private var handled = false

    override fun acknowledge() {
        if (!handled) {
            handled = true
            markPopupContentAcknowledged(this)
        }
    }

    override fun itemAcknowledge(item: AddToListItem) {
        if (!handled) {
            handled = true
            markPopupContentAcknowledged(this)
        }
        markPopupContentItemAcknowledged(this, item)
    }

    override fun failed(message: String) {
        if (!handled) {
            handled = true
            markPopupContentFailed(this, message)
        }
    }

    override fun itemFailed(item: AddToListItem, message: String) {
        markPopupContentItemFailed(this, item, message)
    }

    override fun getSource(): String {
        return AddToListContent.Sources.IN_APP
    }

    override fun getItems(): List<AddToListItem> {
        return items
    }

    override fun hasNoItems(): Boolean {
        return items.isEmpty()
    }

    companion object {
        fun createPopupContent(payloadId: String, items: List<AddToListItem>): PopupContent {
            return PopupContent(payloadId, items)
        }

        fun markPopupContentAcknowledged(content: PopupContent) {
            val params: MutableMap<String, String> = HashMap()
            params[PAYLOAD_ID] = content.payloadId
            EventClient.trackSdkEvent(EventStrings.POPUP_ADDED_TO_LIST, params)
        }

        fun markPopupContentItemAcknowledged(content: PopupContent, item: AddToListItem) {
            val params: MutableMap<String, String> = HashMap()
            params[PAYLOAD_ID] = content.payloadId
            params[TRACKING_ID] = item.trackingId
            params[ITEM_NAME] = item.title
            EventClient.trackSdkEvent(EventStrings.POPUP_ITEM_ADDED_TO_LIST, params)
        }

        fun markPopupContentFailed(content: PopupContent, message: String) {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            EventClient.trackSdkError(EventStrings.POPUP_CONTENT_FAILED, message, eventParams)
        }

        fun markPopupContentItemFailed(
            content: PopupContent,
            item: AddToListItem,
            message: String
        ) {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            eventParams[TRACKING_ID] = item.trackingId
            EventClient.trackSdkError(EventStrings.POPUP_CONTENT_ITEM_FAILED, message, eventParams)
        }

        private const val PAYLOAD_ID = "payload_id"
        private const val TRACKING_ID = "tracking_id"
        private const val ITEM_NAME = "item_name"
    }
}